core/lib/zksync_core/src/consensus/storage.rs

```diff
@@ -0,0 +1,423 @@
+        &mut self,
+        ctx: &ctx::Ctx,
+        start_at: validator::BlockNumber,
+        operator_address: Address,
+    ) -> ctx::Result<validator::FinalBlock> {
+        let mut block = self
+            .fetch_block(ctx, start_at, operator_address)
+            .await
+            .wrap("fetch_block()")?
+            .context("head not found")?;
+        while let Some(next) = self
+            .fetch_block(ctx, block.header.number.next(), operator_address)
+            .await
+            .wrap("fetch_block()")?
+        {
+            block = next;
+        }
+        Ok(block)
+    }
+}
+
+/// Postgres-based [`BlockStore`] implementation, which
+/// considers blocks as stored <=> they have consensus field set.
+#[derive(Debug)]
+pub(super) struct SignedBlockStore {
+    genesis: validator::BlockNumber,
+    head: sync::watch::Sender<validator::BlockNumber>,
+    pool: ConnectionPool,
+    operator_address: Address,
+}
+
+impl SignedBlockStore {
+    /// Creates a new storage handle. `pool` should have multiple connections to work efficiently.
+    pub async fn new(
+        ctx: &ctx::Ctx,
+        pool: ConnectionPool,
+        genesis: &validator::FinalBlock,
+        operator_address: Address,
+    ) -> anyhow::Result<Self> {
+        // Ensure that genesis block has consensus field set in postgres.
+        let head = {
+            let mut storage = storage(ctx, &pool).await.wrap("storage()")?;
+            storage
+                .put_block(ctx, genesis, operator_address)
+                .await
+                .wrap("put_block()")?;
+
+            // Find the last miniblock with consensus field set (aka head).
+            // We assume here that all blocks in range (genesis,head) also have consensus field set.
+            // WARNING: genesis should NEVER be moved to an earlier block.
+            storage
+                .find_head_number(ctx)
+                .await
+                .wrap("find_head_number()")?
+        };
+        Ok(Self {
+            genesis: genesis.header.number,
+            head: sync::watch::channel(head).0,
+            pool,
+            operator_address,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl WriteBlockStore for SignedBlockStore {
+    /// Verify that `payload` is a correct proposal for the block `block_number`.
+    async fn verify_payload(
+        &self,
+        ctx: &ctx::Ctx,
+        block_number: validator::BlockNumber,
+        payload: &validator::Payload,
+    ) -> ctx::Result<()> {
+        tracing::error!("SignedBlockStore::verify_payload({block_number:?})");
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        let want = storage
+            .fetch_payload(ctx, block_number, self.operator_address)
+            .await
+            .wrap("fetch_payload()")?
+            .context("unknown block")?;
+        if payload != &want.encode() {
+            return Err(anyhow::anyhow!("unexpected payload").into());
```

Comment:
Nit: It could make sense to include `payload` / wanted payload into the error message for debugging purposes, although it would make to parse `payload` instead of encoding `want` in this case.

Reply:
done